File: generated_code/cpp/solution_3555_2.cpp
#include <vector>
#include <queue>
#include <utility>

using namespace std;

class Solution {
public:
    vector<int> getFinalState(vector<int>& nums, int k, int multiplier) {
        // Pair structure: {value, original_index}
        // Custom comparator to mimic LeetCode's requirements:
        // 1. Prioritize smaller values.
        // 2. If values are equal, prioritize the smaller index (first occurrence).
        auto comp = [](const pair<int, int>& a, const pair<int, int>& b) {
            if (a.first != b.first) {
                return a.first > b.first; 
            }
            return a.second > b.second;
        };
        
        priority_queue<pair<int, int>, vector<pair<int, int>>, decltype(comp)> minHeap(comp);
        
        // Push all elements with their initial indices into the min-heap
        for (int i = 0; i < nums.size(); ++i) {
            minHeap.push({nums[i], i});
        }
        
        // Perform the k operations
        while (k > 0) {
            auto [val, idx] = minHeap.top();
            minHeap.pop();
            
            // Multiply the minimum element
            val *= multiplier;
            
            // Update the state in the original array and push it back to the heap
            nums[idx] = val;
            minHeap.push({val, idx});
            
            k--;
        }
        
        return nums;
    }
};

File: generated_code/python/solution_3555_2.py
from typing import List
import heapq

class Solution:
    def getFinalState(self, nums: List[int], k: int, multiplier: int) -> List[int]:
        # If multiplier is 1, operations don't change the elements' values
        if multiplier == 1:
            return nums
            
        # Create a min-heap storing tuples of (value, index)
        # This handles finding the minimum value, and breaks ties by the smallest index
        heap = [(val, idx) for idx, val in enumerate(nums)]
        heapq.heapify(heap)
        
        # Perform k operations
        for _ in range(k):
            val, idx = heapq.heappop(heap)
            new_val = val * multiplier
            nums[idx] = new_val
            heapq.heappush(heap, (new_val, idx))
            
        return nums